pam_kwallet.c

```diff
@@ -0,0 +1,823 @@
+            free(path);
+            close(readSaltPipe[1]);
+            exit(-1);
+        }
+
+        struct stat info;
+        if (stat(path, &info) != 0 || info.st_size == 0 || !S_ISREG(info.st_mode)) {
+            syslog(LOG_ERR, "%s: Failed to ensure %s looks like a salt file", logPrefix, path);
+            free(path);
+            close(readSaltPipe[1]);
+            exit(-1);
+        }
+
+        const int fd = open(path, O_RDONLY | O_CLOEXEC);
+        if (fd == -1) {
+            syslog(LOG_ERR, "%s: Couldn't open file: %s because: %d-%s", logPrefix, path, errno, strerror(errno));
+            free(path);
+            close(readSaltPipe[1]);
+            exit(-1);
+        }
+        free(path);
+        char salt[KWALLET_PAM_SALTSIZE] = {};
+        const ssize_t bytesRead = read(fd, salt, KWALLET_PAM_SALTSIZE);
+        close(fd);
+        if (bytesRead != KWALLET_PAM_SALTSIZE) {
+            syslog(LOG_ERR, "%s: Couldn't read the full salt file contents from file. %d:%d", logPrefix, bytesRead, KWALLET_PAM_SALTSIZE);
+            exit(-1);
+        }
+
+        const ssize_t written = better_write(readSaltPipe[1], salt, KWALLET_PAM_SALTSIZE);
+
+        close(readSaltPipe[1]);
+        if (written != KWALLET_PAM_SALTSIZE) {
+            syslog(LOG_ERR, "%s: Couldn't write the full salt file contents to pipe", logPrefix);
+            exit(-1);
+        }
+
+        exit(0);
+    }
+
+    close(readSaltPipe[1]); // we won't be writing from the pipe
+
+    // pam process, just wait for child to finish
+    int status;
+    waitpid(pid, &status, 0);
+    int success = 1;
+    if (status == 0) {
+        const ssize_t readBytes = read(readSaltPipe[0], saltOut, KWALLET_PAM_SALTSIZE);
+        if (readBytes != KWALLET_PAM_SALTSIZE) {
+            pam_syslog(pamh, LOG_ERR, "%s: Couldn't read the full salt file contents from pipe", logPrefix);
+            success = 0;
+        }
+    } else {
+        pam_syslog(pamh, LOG_ERR, "%s: Couldn't read salt file", logPrefix);
+        success = 0;
+    }
+
+    close(readSaltPipe[0]);
+
+    return success;
+}
+
+int kwallet_hash(pam_handle_t *pamh, const char *passphrase, struct passwd *userInfo, char *key)
+{
+    if (!gcry_check_version("1.5.0")) {
+        syslog(LOG_ERR, "%s-kwalletd: libcrypt version is too old", logPrefix);
+        return 1;
+    }
+
+    struct stat info;
+    if (stat(userInfo->pw_dir, &info) != 0 || !S_ISDIR(info.st_mode)) {
+        syslog(LOG_ERR, "%s-kwalletd: user home folder does not exist", logPrefix);
+        return 1;
+    }
+
+    const char *fixpath = "kwalletd/kdewallet.salt";
+    size_t pathSize = strlen(userInfo->pw_dir) + strlen(kdehome) + strlen(fixpath) + 3;//3 == /, / and \0
+    char *path = (char*) malloc(pathSize);
+    sprintf(path, "%s/%s/%s", userInfo->pw_dir, kdehome, fixpath);
+
+    createNewSalt(pamh, path, userInfo);
+
+    char salt[KWALLET_PAM_SALTSIZE] = {};
+    const int readSaltSuccess = readSaltFile(pamh, path, userInfo, salt);
```

Comment:
Here in `readSaltFile()` another sub process is setup that reads the data that has just been created in the previous child process in `createNewSalt()`. This could be more efficient and less redundant if a single child process is setup that create the file *and* returns it contents via a pipe.